File: src/medical-formation.test.js
import test from "node:test";
import assert from "node:assert/strict";
import { initial,step,launchPosition } from "./game.js";
import { startInjury } from "./features.js";
test("healthy players return during medical care without releasing ball or advancing launcher", () => {
 const s=initial(),p=s.players[1];const target=launchPosition(s,1);
 p.x+=3;const before=p.x;
 Object.assign(s.players[7],{x:0,z:0,health:0,actionTime:0});
 startInjury(s,7);const anchor=[s.ball.x,s.ball.z];
 Object.assign(s.players[8],{health:0,actionTime:.4});
 step(s,.04,{x:1,shoot:true},[true,true]);
 assert.ok(p.x<before);assert.equal(s.players[7].stun,1);
 for(let i=0;i<49;i++)step(s,.04,{},[true,true]);
 assert.equal(s.medical.player,7);assert.equal(s.players[8].injury,0);
 assert.deepEqual([p.x,p.z],target);
 assert.deepEqual([s.ball.x,s.ball.z],anchor);assert.equal(s.ball.owner,-1);
 assert.equal(s.time,90);assert.notEqual(s.restartPhase,2);
 assert.equal(s.events.some(e=>e.kind===22||e.kind===23),false);
});

File: server/medical_formation_test.go
package main

import "testing"

func TestFormationDuringMedical(t *testing.T) {
	s := initial()
	p := &s.Players[1]
	x, z := s.launchPosition(1)
	p.X += 3
	before := p.X
	s.Players[7].X = 0
	s.Players[7].Z = 0
	s.Players[7].Health = 0
	s.Players[7].ActionTime = 0
	s.startInjury(7)
	bx, bz := s.Ball.X, s.Ball.Z
	s.Players[8].Health = 0
	s.Players[8].ActionTime = .4
	s.step(.04, [2]Input{})
	if p.X >= before || s.Players[7].Stun != 1 {
		t.Fatal("formation movement or patient changed")
	}
	for i := 0; i < 49; i++ {
		s.step(.04, [2]Input{})
	}
	if s.Medical == nil || s.Medical.Player != 7 || s.Players[8].Injury != 0 {
		t.Fatal("medical serialization")
	}
	if p.X != x || p.Z != z {
		t.Fatal("formation not reached")
	}
	if s.Ball.X != bx || s.Ball.Z != bz || s.Ball.Owner != -1 || s.Time != 90 || s.RestartPhase == 2 {
		t.Fatal("premature restart")
	}
	for _, e := range s.Events[:s.EventCount] {
		if e.Kind == 22 || e.Kind == 23 {
			t.Fatal("launcher during medical")
		}
	}
}
